### ISSUE TITLE
race condition -- DefaultCertificateDiskCache.SaveCertificate

### ISSUE THREAD

**Issue by jgilbert2017** [justcoding121/Titanium-Web-Proxy#816]
Sometimes at startup when I hit a website with multiple simultaneous requests it looks like there is an issue with multiple threads trying to call `SaveCertificate` at the same time.  It doesn't seem to break anything but perhaps there is some synchronization missing around the certificate handling.

```
Exception: System.Exception: Failed to save fake certificate. ---> System.IO.IOException: The process cannot access the file 'C:\crts\thewebsite.com.pfx' because it is being used by another process.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at System.IO.File.InternalWriteAllBytes(String path, Byte[] bytes)
   at System.IO.File.WriteAllBytes(String path, Byte[] bytes)
   at Titanium.Web.Proxy.Network.DefaultCertificateDiskCache.SaveCertificate(String subjectName, X509Certificate2 certificate) in C:\gitsrc\Titanium-Web-Proxy\src\Titanium.Web.Proxy\Certificates\DefaultCertificateDiskCache.cs:line 41
   at Titanium.Web.Proxy.Network.CertificateManager.CreateCertificate(String certificateName, Boolean isRootCertificate) in C:\gitsrc\Titanium-Web-Proxy\src\Titanium.Web.Proxy\Certificates\CertificateManager.cs:line 448
   --- End of inner exception stack trace ---
```

**Comment by giesalabs** [comment]
i'm not a titanium proxy developer, but i saw the code yesterday, i can try to fix it and you can test the changes
and if all works we can send the new code to the project maintainer
do you want to try?

**Comment by honfika** [comment]
Yes, please create a PR

**Comment by giesalabs** [comment]
there are two ways to fix it
1. a complex code that is the fastest in execution
2. a slower code, but a really easy to read code

in the spirit of the rest of the code I wrote a slower code, but really easy to read and maintain
do you agree?

**Comment by giesalabs** [comment]
as example on top of the class:
`private static onlyread object writeLocker = new object();`

and the method:

```
        public void SaveCertificate(string subjectName, X509Certificate2 certificate)
        {
            lock (writeLocker)//wait for other SaveCertificate: is fix for https://github.com/justcoding121/titanium-web-proxy/issues/816
            {
                string filePath = Path.Combine(getCertificatePath(true), subjectName + defaultCertificateFileExtension);
                byte[] exported = certificate.Export(X509ContentType.Pkcs12);

                bool writeError = false;
                int itry = 0;
                do
                {
                    try
                    {
                        File.WriteAllBytes(filePath, exported);
                    }
                    catch (Exception e)
                    {
                        writeError = true;
                        itry++;
                        Thread.Sleep(100);
                    }
                } while (writeError && itry < 30);//max 3 seconds
            }
        }
```

**Comment by jgilbert2017** [comment]
[NOTE: This is no longer an issue for me.  I disabled saving of certificates.  Upon testing, the saved certificates are cleared out upon restart (when initially setting the root certificate).  Assuming this is the intended behavior, I suppose the saved certificates are only used for fast reloading if the certificate is dropped from the in memory cache due to timeout.   For my use case, I'm ok with regenerating the certificate in this case.]

The race in the current code is in `CertificateManager.CreateServerCertificate`. The certificate creation task may be done multiple times due to lack of synchronization across calls to `pendingCertificateCreationTasks.TryGetValue` and `pendingCertificateCreationTasks.TryAdd`.

I've added an async lock (using `SemaphoreSlim`)  for the pending task check/create/add sequence.  This will go far in fixing the issue in that it will prevent 2 of the same cert creation tasks from running simultaneously.  However, there is still a race that allows for potentially duplicate sequential generation (should be very rare) which i think won't cause any issues.
